File: android/app/src/main/kotlin/app/fihaven/ui/theme/Theme.kt
package app.fihaven.ui.theme

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import androidx.activity.ComponentActivity
import androidx.activity.SystemBarStyle
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalView

/// Design tokens ported from client/css/tokens.css (docs/native-contract.md §8).
data class CtColors(
    val bg: Color,
    val surface: Color,
    val surface2: Color,
    val border: Color,
    val text: Color,
    val muted: Color,
    val accent: Color,
    val accentHover: Color,
    val accentBg: Color,
    val green: Color,
    val greenBg: Color,
    val red: Color,
    val redBg: Color,
    val orange: Color,
    val orangeBg: Color,
    val yellow: Color,
    val yellowBg: Color,
    /// Chart series. Deliberately NOT green/red: that pair reads as ΔE 2.1 under
    /// deuteranopia on the dark surface — effectively one color. Blue/orange holds
    /// ΔE >= 26 in both modes. The dark steps are chosen against the dark surface,
    /// not flipped from light (the light steps sit above the dark lightness band).
    val chartIncome: Color,
    val chartSpend: Color,
    /// Plate behind a full-color brand mark. White in both themes on purpose:
    /// these logos are drawn for a white page, and a black wordmark would
    /// disappear if the plate followed the dark surface.
    val logoPlate: Color,
    /// Neutral floor under an issuer tile's brand-colored ring (web
    /// `--chip-edge`). It flips to light on the dark theme because a
    /// near-black tile — Apple's, Bilt's — has no edge of its own there.
    val tileEdge: Color,
)

val LightCt = CtColors(
    bg = Color(0xFFFAFAFB), surface = Color(0xFFFFFFFF), surface2 = Color(0xFFF2F3F6),
    border = Color(0xFFE5E7EB), text = Color(0xFF15161A), muted = Color(0xFF6C6E77),
    accent = Color(0xFF3D6FE1), accentHover = Color(0xFF2F5DCB), accentBg = Color(0xFFEAF0FE),
    green = Color(0xFF15803D), greenBg = Color(0xFFE7F4EC), red = Color(0xFFDC2626), redBg = Color(0xFFFDECEC),
    orange = Color(0xFFC2410C), orangeBg = Color(0xFFFDEEE3), yellow = Color(0xFFA16207), yellowBg = Color(0xFFFBF5DC),
    chartIncome = Color(0xFF3D6FE1), chartSpend = Color(0xFFC2410C),
    logoPlate = Color(0xFFFFFFFF), tileEdge = Color(0x1A000000),
)

val DarkCt = CtColors(
    bg = Color(0xFF0C0D0F), surface = Color(0xFF17181B), surface2 = Color(0xFF1F2126),
    border = Color(0xFF292B31), text = Color(0xFFECEDF0), muted = Color(0xFF868892),
    accent = Color(0xFF6098F6), accentHover = Color(0xFF82AEFA), accentBg = Color(0xFF122544),
    green = Color(0xFF34C57B), greenBg = Color(0xFF0E2B1A), red = Color(0xFFF87171), redBg = Color(0xFF2B1414),
    orange = Color(0xFFFB923C), orangeBg = Color(0xFF2B1A0C), yellow = Color(0xFFFBBF24), yellowBg = Color(0xFF2B2008),
    chartIncome = Color(0xFF4A87EE), chartSpend = Color(0xFFD9700F),
    logoPlate = Color(0xFFFFFFFF), tileEdge = Color(0x2BFFFFFF),
)

val LocalCt = staticCompositionLocalOf { LightCt }

/// Convenience accessor: `Ct.colors`.
object Ct {
    val colors: CtColors
        @Composable @ReadOnlyComposable get() = LocalCt.current
}

@Composable
fun FiHavenTheme(pref: ThemePref = ThemePref.SYSTEM, content: @Composable () -> Unit) {
    val dark = when (pref) {
        ThemePref.SYSTEM -> isSystemInDarkTheme()
        ThemePref.LIGHT -> false
        ThemePref.DARK -> true
    }
    val ct = if (dark) DarkCt else LightCt
    // Re-apply edge-to-edge whenever `dark` flips. enableEdgeToEdge()'s default
    // styles decide the system bars' icon color and scrim from the *system*
    // dark-mode setting, but `dark` above is the user's in-app preference, which
    // is free to disagree with it — so a Dark app on a Light phone drew dark
    // status-bar icons onto our dark background, and vice versa. Handing the
    // styles an explicit detectDarkMode is the supported way to bind them to
    // something other than uiMode. MainActivity's bare call covers the frames
    // before this runs.
    val view = LocalView.current
    if (!view.isInEditMode) {
        val activity = view.context.findActivity() as? ComponentActivity
        DisposableEffect(activity, dark) {
            activity?.enableEdgeToEdge(
                statusBarStyle = SystemBarStyle.auto(TRANSPARENT, TRANSPARENT) { dark },
                navigationBarStyle = SystemBarStyle.auto(NavBarLightScrim, NavBarDarkScrim) { dark },
            )
            onDispose {}
        }
    }
    val scheme = if (dark) {
        darkColorScheme(
            primary = ct.accent, background = ct.bg, surface = ct.surface,
            onPrimary = Color.White, onBackground = ct.text, onSurface = ct.text,
        )
    } else {
        lightColorScheme(
            primary = ct.accent, background = ct.bg, surface = ct.surface,
            onPrimary = Color.White, onBackground = ct.text, onSurface = ct.text,
        )
    }
    CompositionLocalProvider(LocalCt provides ct) {
        MaterialTheme(colorScheme = scheme, typography = fihavenTypography()) {
            // Make Manrope the default for every Text() that doesn't set its own.
            CompositionLocalProvider(
                LocalTextStyle provides LocalTextStyle.current.copy(fontFamily = Manrope),
                content = content,
            )
        }
    }
}

/// A composable's context is not always the Activity — Compose wraps it, and
/// inside a Dialog it is the dialog's own themed context — so walk the wrappers
/// rather than casting.
private tailrec fun Context.findActivity(): Activity? = when (this) {
    is Activity -> this
    is ContextWrapper -> baseContext.findActivity()
    else -> null
}

private const val TRANSPARENT = 0

/// Navigation-bar fill below API 29, which has no contrast enforcement to keep
/// the buttons legible over app content (and on API 26, no light-button mode
/// either). Copied from androidx.activity's own defaults so the bar looks
/// identical to what plain enableEdgeToEdge() produced — only which of the two
/// gets picked changes. Both are unused from API 29 up: there the bar is simply
/// transparent.
private const val NavBarLightScrim = 0xE6FFFFFF.toInt()
private const val NavBarDarkScrim = 0x801B1B1B.toInt()

File: ios/FiHavenApp/Sources/Theme/Theme.swift
import SwiftUI
import FiHavenCore
#if canImport(UIKit)
import UIKit
#endif

/// Design tokens ported from client/css/tokens.css. Each color resolves
/// light/dark from the system appearance, matching the web's
/// `:root` / `[data-theme="dark"]` palettes (docs/native-contract.md §8).
enum Theme {
    // ── Palette ──────────────────────────────────────────────────────
    static let bg          = dyn(0xFAFAFB, 0x0C0D0F)
    static let surface     = dyn(0xFFFFFF, 0x17181B)
    static let surface2    = dyn(0xF2F3F6, 0x1F2126)
    static let border      = dyn(0xE5E7EB, 0x292B31)
    static let text        = dyn(0x15161A, 0xECEDF0)
    static let muted       = dyn(0x555661, 0x9496A3)
    static let accent      = dyn(0x3D6FE1, 0x6098F6)
    static let accentHover = dyn(0x2F5DCB, 0x82AEFA)
    static let accentBg    = dyn(0xEAF0FE, 0x122544)
    static let green       = dyn(0x15803D, 0x34C57B)
    static let greenBg     = dyn(0xE7F4EC, 0x0E2B1A)
    static let red         = dyn(0xDC2626, 0xF87171)
    static let redBg       = dyn(0xFDECEC, 0x2B1414)
    static let orange      = dyn(0xC2410C, 0xFB923C)
    static let orangeBg    = dyn(0xFDEEE3, 0x2B1A0C)
    static let yellow      = dyn(0xA16207, 0xFBBF24)
    static let yellowBg    = dyn(0xFBF5DC, 0x2B2008)
    /// Chart series. Deliberately NOT green/red: that pair reads as ΔE 2.1 under
    /// deuteranopia on the dark surface — effectively one color. Blue/orange holds
    /// ΔE ≥ 26 in both modes. The dark steps are chosen against the dark surface,
    /// not flipped from light (the light steps sit above the dark lightness band).
    static let chartIncome = dyn(0x3D6FE1, 0x4A87EE)
    static let chartSpend  = dyn(0xC2410C, 0xD9700F)
    /// Plate behind a full-color brand mark. White in both themes on purpose:
    /// these logos are drawn for a white page, and a black wordmark would
    /// disappear if the plate followed the dark surface.
    static let logoPlate   = dyn(0xFFFFFF, 0xFFFFFF)

    // ── Radii (px → pt) ──────────────────────────────────────────────
    static let radius: CGFloat = 10
    static let radiusCard: CGFloat = 14
    static let radiusPill: CGFloat = 11

    // ── Type ─────────────────────────────────────────────────────────
    // Bundled OFL fonts: Manrope (variable wght axis) for UI, IBM Plex
    // Mono for numbers. Manrope's default instance is ExtraLight, so we
    // always pin an explicit weight. `relativeTo` ties each size to a
    // Dynamic Type text style so Larger Text scales across the app.
    static func title(_ size: CGFloat, relativeTo style: Font.TextStyle = .title) -> Font {
        .custom("Manrope", size: size, relativeTo: style).weight(.heavy)
    }
    static func ui(
        _ size: CGFloat,
        weight: Font.Weight = .regular,
        relativeTo style: Font.TextStyle = .body
    ) -> Font {
        .custom("Manrope", size: size, relativeTo: style).weight(weight)
    }
    static func mono(
        _ size: CGFloat,
        weight: Font.Weight = .regular,
        relativeTo style: Font.TextStyle = .body
    ) -> Font {
        // Two static faces bundled (Regular + Medium); map heavier UI
        // weights onto Medium.
        .custom(
            weight == .regular ? "IBMPlexMono-Regular" : "IBMPlexMono-Medium",
            size: size,
            relativeTo: style
        )
    }

    // ── Color resolution ─────────────────────────────────────────────
    static func dyn(_ light: UInt, _ dark: UInt) -> Color {
        #if canImport(UIKit)
        return Color(UIColor { traits in
            uiColor(traits.userInterfaceStyle == .dark ? dark : light)
        })
        #else
        return rgb(light)
        #endif
    }

    #if canImport(UIKit)
    private static func uiColor(_ hex: UInt) -> UIColor {
        UIColor(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: 1
        )
    }
    #endif

    /// An issuer brand color, nudged toward legibility against the card
    /// surface it's drawn on (see `BrandColor`) — Visa's navy and Apple's
    /// black are otherwise invisible in dark mode.
    static func brand(_ color: UInt32) -> Color {
        #if canImport(UIKit)
        return Color(UIColor { traits in
            let surface: UInt32 = traits.userInterfaceStyle == .dark ? 0x17181B : 0xFFFFFF
            return uiColor(UInt(BrandColor.legible(color, on: surface)))
        })
        #else
        return rgb(UInt(BrandColor.legible(color, on: 0xFFFFFF)))
        #endif
    }

    /// Background for a monogram chip: dark enough for white initials, and
    /// in dark mode still distinguishable from the card behind it (an almost
    /// black brand color would otherwise vanish into the surface).
    static func chip(_ color: UInt32) -> Color {
        let readable = BrandColor.legible(color, on: 0xFFFFFF)
        #if canImport(UIKit)
        return Color(UIColor { traits in
            let surface: UInt32 = traits.userInterfaceStyle == .dark ? 0x17181B : 0xFFFFFF
            return uiColor(UInt(BrandColor.legible(readable, on: surface, minContrast: 1.6)))
        })
        #else
        return rgb(UInt(readable))
        #endif
    }

    /// A packed 0xRRGGBB color, exactly as authored — for the layers of a
    /// full-color brand mark, which must not be nudged for contrast.
    static func exact(_ color: UInt32) -> Color { rgb(UInt(color)) }

    private static func rgb(_ hex: UInt) -> Color {
        Color(
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255
        )
    }
}
